### PR TITLE
MIR-942 possibility to allow attributes in abstract

### PR DESCRIPTION
**Pull request by sebhofmann**
[Link to jira](https://mycore.atlassian.net/browse/MIR-924).

### CODE REVIEW

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        final Element test3 = body.getElementById("test3");
+        final Element test4 = body.getElementById("test4");
+        final Element test5 = body.getElementById("test5");
+        final Elements p = body.getElementsByTag("p");
+        final Elements i = body.getElementsByTag("i");
+
+        Assert.assertEquals("Element test1 should be present", "h1", test1.tagName());
+
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:48]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

**Comment by yagee-de** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 17
           

Complexity increasing per file
==============================
- mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java  1
         

Complexity decreasing per file
==============================
+ mir-module/src/main/java/org/mycore/mir/authorization/MIRStrategy.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128&bid=20283755)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
+        Assert.assertFalse("If an attribute is href then only http and https is allowed", test4.hasAttr("href"));
+
+        Assert.assertEquals("Element test5 should be present", "a", test5.tagName());
+        Assert.assertFalse("Onmouseover is not allowed", test5.hasAttr("onmouseover"));
+
+        Assert.assertEquals("Element p should be present", 1, p.size());
+        Assert.assertEquals("Element i should be present", 1, i.size());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:60]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package org.mycore.mir.editor;
+
+import org.apache.logging.log4j.LogManager;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.safety.Whitelist;
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:7]: Codacy found an issue: [Avoid unused imports such as 'org.jsoup.safety.Whitelist'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        final Element test2 = body.getElementById("test2");
+        final Element test3 = body.getElementById("test3");
+        final Element test4 = body.getElementById("test4");
+        final Element test5 = body.getElementById("test5");
+        final Elements p = body.getElementsByTag("p");
+        final Elements i = body.getElementsByTag("i");
+
+        Assert.assertEquals("Element test1 should be present", "h1", test1.tagName());
+
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:47]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import org.jsoup.nodes.Element;
+import org.jsoup.safety.Whitelist;
+import org.jsoup.select.Elements;
+import org.junit.Assert;
+import org.junit.Test;
+import org.mycore.common.MCRTestCase;
+
+import java.util.Map;
+
+import static org.junit.Assert.*;
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:15]: Codacy found an issue: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
+        Assert.assertTrue("Attribute title is allowed", test2.hasAttr("title"));
+
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
+        Assert.assertFalse("If an attribute is href then only http and https is allowed", test4.hasAttr("href"));
+
+        Assert.assertEquals("Element test5 should be present", "a", test5.tagName());
+        Assert.assertFalse("Onmouseover is not allowed", test5.hasAttr("onmouseover"));
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:57]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertFalse' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java (hunk 5 above)
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:15]: Codacy found an issue: [Using the '.*' form of import should be avoided - org.junit.Assert.*.](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
+        Assert.assertTrue("Attribute title is allowed", test2.hasAttr("title"));
+
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
+        Assert.assertFalse("If an attribute is href then only http and https is allowed", test4.hasAttr("href"));
+
+        Assert.assertEquals("Element test5 should be present", "a", test5.tagName());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:56]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package org.mycore.mir.editor;
+
+import org.apache.logging.log4j.LogManager;
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:3]: Codacy found an issue: [Avoid unused imports such as 'org.apache.logging.log4j.LogManager'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java (hunk 3 above)
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:7]: Codacy found an issue: [Unused import - org.jsoup.safety.Whitelist.](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
+        Assert.assertFalse("If an attribute is href then only http and https is allowed", test4.hasAttr("href"));
+
+        Assert.assertEquals("Element test5 should be present", "a", test5.tagName());
+        Assert.assertFalse("Onmouseover is not allowed", test5.hasAttr("onmouseover"));
+
+        Assert.assertEquals("Element p should be present", 1, p.size());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:59]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        Assert.assertEquals("Element test1 should be present", "h1", test1.tagName());
+
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
+        Assert.assertTrue("Attribute title is allowed", test2.hasAttr("title"));
+
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
+        Assert.assertFalse("If an attribute is href then only http and https is allowed", test4.hasAttr("href"));
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:54]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertFalse' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        final Elements p = body.getElementsByTag("p");
+        final Elements i = body.getElementsByTag("i");
+
+        Assert.assertEquals("Element test1 should be present", "h1", test1.tagName());
+
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
+        Assert.assertTrue("Attribute title is allowed", test2.hasAttr("title"));
+
+        Assert.assertNull("Element test3 should not be present", test3);
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:51]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertNull' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        final Element body = cleanDocument.body();
+        final Element test1 = body.getElementById("test1");
+        final Element test2 = body.getElementById("test2");
+        final Element test3 = body.getElementById("test3");
+        final Element test4 = body.getElementById("test4");
+        final Element test5 = body.getElementById("test5");
+        final Elements p = body.getElementsByTag("p");
+        final Elements i = body.getElementsByTag("i");
+
+        Assert.assertEquals("Element test1 should be present", "h1", test1.tagName());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:45]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java (hunk 8 above)
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:3]: Codacy found an issue: [Unused import - org.apache.logging.log4j.LogManager.](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        Assert.assertEquals("Element test2 should be present", "a", test2.tagName());
+        Assert.assertTrue("Attribute href with http is allowed", test2.hasAttr("href"));
+        Assert.assertTrue("Attribute title is allowed", test2.hasAttr("title"));
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:49]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

File: mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+        Assert.assertNull("Element test3 should not be present", test3);
+
+        Assert.assertEquals("Element test4 should be present", "a", test4.tagName());
```
Comment (yagee-de) [inline mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java:53]: Codacy found an issue: [Unnecessary use of fully qualified name 'Assert.assertEquals' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128)

**Comment by yagee-de** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java  1
         

Complexity decreasing per file
==============================
+ mir-module/src/main/java/org/mycore/mir/authorization/MIRStrategy.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128&bid=20283755)

**Comment by fluetze** [comment]
There are changes in MIRStrategy and mods2metsmods which are not related to this issue.

**Comment by sebhofmann** [comment]
Branch is not testable without the changes in MIR-Strategy

**Comment by yagee-de** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java  1
         

Complexity decreasing per file
==============================
+ mir-module/src/main/java/org/mycore/mir/authorization/MIRStrategy.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128&bid=20283755)

**Comment by sebhofmann** [comment]
> Branch is not testable without the changes in MIR-Strategy

Changes can be removed if https://github.com/MyCoRe-Org/mir/pull/532 is merged and this branch rebased

**Review by kkrebs — APPROVED** [review]
works with this configuration:
```
MIR.Editor.HTML.Elements=h1[id];a[id,href,title];p br i strong em u s sub sup table thead tr th tbody td hr ol ul li blockquote pre address h2 h3 h4 h5 h6 div q big small code del
MIR.Editor.HTML.Elements=%MIR.Editor.HTML.Elements% %MIR.Editor.MathML.Elements%
```

**Comment by yagee-de** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- mir-module/src/test/java/org/mycore/mir/editor/MIREditorUtilsTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mir/pullRequest?prid=6223128&bid=20283755)

**Pull request merged**